napari/plugins/utils.py

```diff
@@ -9,14 +12,104 @@
 from . import _npe2, plugin_manager
 
 
-def get_preferred_reader(_path):
-    """Return preferred reader for _path from settings, if one exists."""
+class MatchFlag(IntFlag):
+    NONE = 0
+    SET = 1
+    ANY = 2
+    STAR = 4
+
+
+@lru_cache
+def score_specificity(pattern: str) -> Tuple[bool, int, List[MatchFlag]]:
+    """Score an fnmatch pattern, with higher specificities having lower scores.
+
+    Absolute paths have highest specificity,
+    followed by paths with the most nesting,
+    then by path segments with the least ambiguity.
+
+    Parameters
+    ----------
+    pattern : str
+        Pattern to score.
+
+    Returns
+    -------
+    relpath : boolean
+        Whether the path is relative or absolute.
+    nestedness : negative int
+        Level of nestedness of the path, lower is deeper.
+    score : List[MatchFlag]
+        Path segments scored by ambiguity, higher score is higher ambiguity.
+    """
+    pattern = osp.normpath(pattern)
+
+    segments = pattern.split(osp.sep)
+    score: List[MatchFlag] = []
+    ends_with_star = False
+
+    def add(match_flag):
+        score[-1] |= match_flag
+
+    # built-in fnmatch does not allow you to escape meta-characters
+    # so we don't need to handle them :)
+    for segment in segments:
+        # collapse foo/*/*/*.bar or foo*/*.bar but not foo*bar/*.baz
+        if segment and not (ends_with_star and segment.startswith('*')):
+            score.append(MatchFlag.NONE)
```

Comment:
tried this
```python
for segment in segments:
        if not segment or ends_with_star and segment.startswith('*'):
            # collapse foo/*/*/*.bar or foo*/*.bar but not foo*bar/*.baz
            continue
        
        segment_score = MatchFlag.NONE

        if '*' in segment:
            segment_score |= MatchFlag.STAR
        if '?' in segment:
            segment_score |= MatchFlag.ANY
        if '[' in segment and ']' in segment[segment.index('[') :]:
            segment_score |= MatchFlag.SET

        score.append(segment_score)

        ends_with_star = segment.endswith('*')
```

but realized that it actually failed an edge case: on `*/*?.tif` it will rate it as `Match.STAR` when it should actually collapse to `MatchFlag.STAR | MatchFlag.ANY`. capturing this in the tests now

Reply:
ok!

Reply:
maybe that was a bad example since `*?` is equal to `*`, a better example would be `*/*.t?f`